# Lab 06 - EDA - Annette Bazan

**What is EDA (Exploratory Data Analysis)?**
1. EDA is a process in data analytics use to analyze and investigate datasets and find out the main characteristics.
2. EDA helps data analysts to discover patterns in the dataset and examine their assumptions.
3. EDA is also used to find out about the relationship between the datasets variables.
4. It was developed by the American mathametician, John Turkey in the 1970s.
5. EDA helps analysts to answer the management team questions.
6. When EDA is done,  the dataset can be used for more sophisticated applications of AI such as ML.

**Types of EDA**
1. Univariate: Examining one variables
2. Bivariate: Examining two variables
3. Multivariate: Examining more than two variables

**EDA Tools:**
1. Graphical EDA (data visualization): plots, graphs, charts
2. Non-Graphical EDA: Examining the relationship between the dataset variables.

**EDA Most Popular Languages:**
1. Python
2. R

### Steps in completing a Data Analysis Project:
1. Introduction:
   - Context: (background of dataset)
   - Objectives: (why we are doing this?)
   - Data description:(knowing the name of each variable and a short description of that variable)
2. Importing the needed libraries
3. Loading the dataset
4. EDA (univariate, bivariate, multivariate)
   - Basic data exploration
   - Advanced data exploration
   - Data Cleaning
   - Data Engineering
   - Summary of EDA
5. Data Visualization
6. Data Mining
   - either/or: if the data analysis is a descriptive one, data mining and answering the data mining questions.
   - If the data analysis is a predictive one: data will be used in a more advanced step of machine learning or any model-building process.
7. Data Reporting (presentaion)

### 1. Introduction: To be completed by you
- Context:
- Objective:
- Data Description:

### 2. Importing libraries

In [1]:
# let's import the libraries
import numpy as np
import pandas as pd

### 3. Loading (Reading) Dataset

In [2]:
# let's read the dataset (load)
df = pd.read_csv('movies.csv')

In [3]:
# creating a copy of the dataset as a backup
movies = df.copy()

### 4. EDA

#### Basic data exploration
**Make sure the Dataset has been loaded correctly**

**Data Attributes:** To be completed by you
1. shape: (782,5)
2. size: 3910
3. ndim: d
4. columns: Index(['Rank', 'Title', 'Studio', 'Gross', 'Year'], dtype='object')
5. dtypes: Rank       int64
           Title     object
           Studio    object
           Gross     object
           Year       int64
           dtype: object

In [4]:
# Lets ensure the dataset has been loaded correctly
# The 1st 5 rows of the dataset
movies.head()

Rank                         Title       Studio       Gross  Year
0     1             Avengers: Endgame  Buena Vista  $2,796.30   2019
1     2                        Avatar          Fox  $2,789.70   2009
2     3                       Titanic    Paramount  $2,187.50   1997
3     4  Star Wars: The Force Awakens  Buena Vista  $2,068.20   2015
4     5        Avengers: Infinity War  Buena Vista  $2,048.40   2018

In [5]:
# let's show the last 5 rows of the dataset
movies.tail()

Rank                     Title           Studio     Gross  Year
777   778                 Yogi Bear  Warner Brothers  $201.60   2010
778   779       Garfield: The Movie              Fox  $200.80   2004
779   780               Cats & Dogs  Warner Brothers  $200.70   2001
780   781  The Hunt for Red October        Paramount  $200.50   1990
781   782                  Valkyrie              MGM  $200.30   2008

In [6]:
movies.shape

(782, 5)

In [7]:
movies.size

3910

In [8]:
movies.ndim

2

In [9]:
movies.columns

Index(['Rank', 'Title', 'Studio', 'Gross', 'Year'], dtype='object')

In [10]:
movies.dtypes

Rank       int64
Title     object
Studio    object
Gross     object
Year       int64
dtype: object

### Advanced Data Exploration

In [11]:
# Let's get more information about the dataset
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 782 entries, 0 to 781
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Rank    782 non-null    int64 
 1   Title   782 non-null    object
 2   Studio  782 non-null    object
 3   Gross   782 non-null    object
 4   Year    782 non-null    int64 
dtypes: int64(2), object(3)
memory usage: 30.7+ KB


**Observations:** 
1. The dataset consists of 782 rows and 5 columns, including Rank, Title, Studio, Gross, and Year.
2. All columns have no missing values, with each containing 782 non-null entries.
3. The Rank and Year columns are of integer type, while Title, Studio, and Gross are of object type (strings).
4. The Gross column, although representing monetary values, is stored as a string, which may require conversion for numerical analysis.
5. The DataFrame provides details on movies or shows, including their rank, title, studio, gross earnings, and release year, spanning a range of 782 entries.

In [12]:
# Let's examine the statistical summary of the dataset (to transpose .T)
movies.describe() .T

count         mean         std     min      25%     50%      75%     max
Rank  782.0   391.500000  225.888247     1.0   196.25   391.5   586.75   782.0
Year  782.0  2006.620205   10.026227  1939.0  2001.00  2009.0  2014.00  2019.0

**Observations:**
1. The movies start from 1939 all the way to 2019
2. The mean for the movies dataset is 2006
3. The median is 2009
4. It shows Rank and Year and the total count 782.
5. The 25% shows 2001 and 75% shows 2014

In [13]:
# Let's get the summary of the object variables
movies.describe(include = 'object')

Title           Studio     Gross
count                    782              782       782
unique                   773               37       701
top     Beauty and the Beast  Warner Brothers  $225.90 
freq                       2              132         3

**Observations:**
1. No missing values
2. There are 37 studio in this dataset
3. The top title is Beauty and the Beast with a Gross of $225.90, and There are two versions of Beauty and the Beast.
4. It is a descriptive dataset it has past data
5. Warner Brothers studio will be seen 132 times.

In [14]:
# Let's assign our own index by using the Title variable(index is not part of the dataset)
movies.set_index('Title', inplace = True)

In [15]:
movies.head()

Rank       Studio       Gross  Year
Title                                                            
Avengers: Endgame                1  Buena Vista  $2,796.30   2019
Avatar                           2          Fox  $2,789.70   2009
Titanic                          3    Paramount  $2,187.50   1997
Star Wars: The Force Awakens     4  Buena Vista  $2,068.20   2015
Avengers: Infinity War           5  Buena Vista  $2,048.40   2018

In [16]:
movies.shape

(782, 4)

In [17]:
# Let's sort our index
movies.sort_index().head()

Rank           Studio     Gross  Year
Title                                                  
10,000 B.C.        536  Warner Brothers  $269.80   2008
101 Dalmatians     708      Buena Vista  $215.90   1961
101 Dalmatians     425      Buena Vista  $320.70   1996
2 Fast 2 Furious   632        Universal  $236.40   2003
2012                93             Sony  $769.70   2009

In [18]:
# Sorting the dataset by values ascending
movies.sort_values('Year').head()

Rank       Studio     Gross  Year
Title                                                
Gone with the Wind   288          MGM  $402.40   1939
Bambi                540          RKO  $267.40   1942
101 Dalmatians       708  Buena Vista  $215.90   1961
The Jungle Book      755  Buena Vista  $205.80   1967
The Godfather        604    Paramount  $245.10   1972

In [19]:
# Sorting the dataset by values descending
movies.sort_values('Year').tail()

Rank           Studio       Gross  Year
Title                                                                     
Men in Black International          686             Sony    $220.80   2019
John Wick: Chapter 3 - Parabellum   458        Lionsgate    $304.70   2019
Pokemon Detective Pikachu           263  Warner Brothers    $427.50   2019
Dark Phoenix                        603              Fox    $245.10   2019
Avengers: Endgame                     1      Buena Vista  $2,796.30   2019

**Counting the (Unique)Values**
- Using the value_counts method.
* There are two handy methods:
1. Sort_values (sort values)
2. value_counts (display the sorted unique values)

In [20]:
# Let's look at the unique values in the studio variable
studios = movies['Studio'].value_counts()
studios

Studio
Warner Brothers           132
Buena Vista               125
Fox                       117
Universal                 109
Sony                       86
Paramount                  76
Dreamworks                 27
Lionsgate                  21
New Line                   16
MGM                        11
TriStar                    11
Miramax                    10
Weinstein                   6
Columbia                    5
WGUSA                       4
Polygram                    2
Orion                       2
SonR                        2
Dimension                   2
Vestron                     1
USA                         1
Lions                       1
Focus                       1
Rela.                       1
CL                          1
Pathe                       1
Artisan                     1
IFC                         1
GrtIndia                    1
RKO                         1
UTV                         1
FUN                         1
FR                          1
New

In [21]:
# len() tells you how many variables
len(studios)

37

In [22]:
# Display the top 5 studios
studios = movies['Studio'].value_counts().head()
studios

Studio
Warner Brothers    132
Buena Vista        125
Fox                117
Universal          109
Sony                86
Name: count, dtype: int64

In [23]:
type(studios)

pandas.core.series.Series

### **Observations:**
* Warner Brothers have the most movies in this dataset
* 18 Studios have 1 movie in this dataset
* 4 studios have 2 movies in this dataset

### Data Mining Questions

In [24]:
# 1. What are the top 5 studios in this dataset?
studios.head()

Studio
Warner Brothers    132
Buena Vista        125
Fox                117
Universal          109
Sony                86
Name: count, dtype: int64

### Data Indexers
- There are 2 types of indexers:
 1. iloc indexer -loc=location: This is implicit indexing (not clearly described/stated). It is integer-based and refers to the position of the element. It starts with 0.
 2. loc indexer: This is explicit indexing (clearly stated/described). It is the user-defined name and acts as a locator of the element.

In [29]:
# 2. What is the movie sitting in position 300?
movies.iloc[300]

Rank           301
Studio         Fox
Gross     $389.70 
Year          2016
Name: Independence Day: Resurgence, dtype: object

In [32]:
# 3. What is the information about the movie"Forrest Gump"?
movies.loc['Forrest Gump']

Rank            119
Studio    Paramount
Gross      $677.90 
Year           1994
Name: Forrest Gump, dtype: object

In [35]:
# 4. Is there only one version of the movie "101 Dalmatians"?
movies.loc["101 Dalmatians"]

Rank       Studio     Gross  Year
Title                                            
101 Dalmatians   425  Buena Vista  $320.70   1996
101 Dalmatians   708  Buena Vista  $215.90   1961

In [43]:
# 5. What are the movies created by Universal Studio?
# Let's do double indexing
universal = movies[movies['Studio']=='Universal'] 
universal

Rank     Studio       Gross  Year
Title                                                            
Jurassic World                     6  Universal  $1,671.70   2015
Furious 7                          8  Universal  $1,516.00   2015
Jurassic World: Fallen Kingdom    13  Universal  $1,309.50   2018
The Fate of the Furious           17  Universal  $1,236.00   2017
Minions                           19  Universal  $1,159.40   2015
...                              ...        ...         ...   ...
The Break-Up                     763  Universal    $205.00   2006
Everest                          766  Universal    $203.40   2015
Patch Adams                      772  Universal    $202.30   1998
Kindergarten Cop                 775  Universal    $202.00   1990
Straight Outta Compton           776  Universal    $201.60   2015

[109 rows x 4 columns]

In [44]:
# 6. What are the top 5 movies by Universal Studio?
universal.head()

Rank     Studio       Gross  Year
Title                                                            
Jurassic World                     6  Universal  $1,671.70   2015
Furious 7                          8  Universal  $1,516.00   2015
Jurassic World: Fallen Kingdom    13  Universal  $1,309.50   2018
The Fate of the Furious           17  Universal  $1,236.00   2017
Minions                           19  Universal  $1,159.40   2015

In [49]:
# 7. What are the movies released in 2015?
released_2015 = movies[movies['Year'] == 2015]
released_2015

Rank           Studio       Gross  \
Title                                                                         
Star Wars: The Force Awakens                 4      Buena Vista  $2,068.20    
Jurassic World                               6        Universal  $1,671.70    
Furious 7                                    8        Universal  $1,516.00    
Avengers: Age of Ultron                      9      Buena Vista  $1,405.40    
Minions                                     19        Universal  $1,159.40    
Spectre                                     58             Sony    $880.70    
Inside Out                                  69      Buena Vista    $857.60    
Mission: Impossible - Rogue Nation         118        Paramount    $682.70    
The Hunger Games: Mockingjay - Part 2      130        Lionsgate    $653.40    
The Martian                                137              Fox    $630.20    
Fifty Shades of Grey                       165        Universal    $571.00    
Cinderella                                 180      Buena Vista    $543.50    
The Revenant                               186              Fox    $533.00    
Ant-Man                                    200      Buena Vista    $519.30    
Hotel Transylvania 2                       224             Sony    $474.80    
San Andreas                                226  Warner Brothers    $474.00    
Terminator: Genisys                        254        Paramount    $440.60    
Kingsman: The Secret Service               275              Fox    $414.40    
Home                                       302              Fox    $386.00    
Mad Max: Fury Road                         314  Warner Brothers    $378.90    
The Good Dinosaur                          405      Buena Vista    $332.20    
Taken 3                                    416              Fox    $326.50    
The SpongeBob Movie: Sponge Out of Water   419        Paramount    $325.20    
Maze Runner: The Scorch Trials             438              Fox    $312.30    
The Divergent Series: Insurgent            478        Lionsgate    $297.30    
Pitch Perfect 2                            504        Universal    $287.50    
Paddington                                 539        Dimension    $268.00    
Mojin: The Lost Legend                     560            WGUSA    $259.40    
Lost in Hong Kong                          567            WGUSA    $256.30    
The Peanuts Movie                          595              Fox    $246.20    
Pixels                                     605             Sony    $244.90    
Daddy's Home                               613        Paramount    $242.80    
Spy                                        637              Fox    $235.70    
Alvin and the Chipmunks: The Road Chip     641              Fox    $234.80    
Goodbye Mr. Loser                          658               CL    $228.10    
Ted 2                                      702        Universal    $216.70    
Tomorrowland                               745      Buena Vista    $209.20    
Everest                                    766        Universal    $203.40    
Straight Outta Compton                     776        Universal    $201.60    

                                          Year  
Title                                           
Star Wars: The Force Awakens              2015  
Jurassic World                            2015  
Furious 7                                 2015  
Avengers: Age of Ultron                   2015  
Minions                                   2015  
Spectre                                   2015  
Inside Out                                2015  
Mission: Impossible - Rogue Nation        2015  
The Hunger Games: Mockingjay - Part 2     2015  
The Martian                               2015  
Fifty Shades of Grey                      2015  
Cinderella                                2015  
The Revenant                              2015  
Ant-Man                                   2015  
Hotel Transylvania

In [51]:
len(released_2015)

39

In [52]:
# 8. What are the top 5 movies in 2015?
released_2015.head()

Rank       Studio       Gross  Year
Title                                                            
Star Wars: The Force Awakens     4  Buena Vista  $2,068.20   2015
Jurassic World                   6    Universal  $1,671.70   2015
Furious 7                        8    Universal  $1,516.00   2015
Avengers: Age of Ultron          9  Buena Vista  $1,405.40   2015
Minions                         19    Universal  $1,159.40   2015

In [62]:
# 9. What are the movies released by Fox Studio in 2006?
fox_2006 = movies[(movies['Studio'] == 'Fox') & (movies['Year'] == 2006)]
fox_2006

Rank Studio     Gross  \
Title                                                                       
Ice Age: The Meltdown                                125    Fox  $660.90    
Night at the Museum                                  164    Fox  $574.50    
X-Men: The Last Stand                                238    Fox  $459.40    
The Devil Wears Prada                                415    Fox  $326.60    
Borat: Cultural Learnings of America for Make B...   554    Fox  $261.60    
Eragon                                               586    Fox  $249.50    

                                                    Year  
Title                                                     
Ice Age: The Meltdown                               2006  
Night at the Museum                                 2006  
X-Men: The Last Stand                               2006  
The Devil Wears Prada                               2006  
Borat: Cultural Learnings of America for Make B...  2006  
Eragon                                              2006

In [63]:
# 10. What are the top 3 movies released by Fox in 2006?
fox_2006.head(3)

Rank Studio     Gross  Year
Title                                             
Ice Age: The Meltdown   125    Fox  $660.90   2006
Night at the Museum     164    Fox  $574.50   2006
X-Men: The Last Stand   238    Fox  $459.40   2006

### **Observations:**
* Forrest Gump is from the studio Paramount
* The rank of this movie is 119
* It grossed $677.90
* Came out 1994
* In position 300 is a movie made by Fox studio

#### Conclusion: 27
1. when looking for the top movies use .head() if its a certain number like top 4 you use it as .head(4)
2. The top 3 movies from Fox released by 2006 is Ice Age: The Meltdown, Night at the Musuem and X-Men: The Last Stand.
3. The highest gross was Ice Age: The Meltdown which grossed $660.90.
4. The lowest gross was Eragon with the gross $249.50 from Fox.
5. The top 5 movies released 2015 are Star Wars: The Force Awakens, Jurassic World, Furious 7, Avengers: Age of Ultron and Minions.
6. For 2015 the top movie grossed $2,068.20 and was from the Buena Vista Studio.
7. 39 movies were released in 2015
8. Buena vista was the top studio and Universal was the bottom studio of movies released in 2015.
9. It is rank 301 becuase the pandas built in index starts with 0 when looking at a movie in poisition 300.
10. The movie in this position 300 is Independence Day: Resurgence that came out in 2016
11. There was 2 movies of 101 Dalmatians in this dataset.
12. Top 5 studios in this dataset are :Warner Brothers, Buena Vista, Fox, Universal and Sony.
13. Data mining questions are combing through the dataset and finding out specific information to help understand it.
14. There are 37 studios in total in this dataset.
15. Warner brothers has the most movies in this dataset.
16. Miramax has 10 movies in this dataset.
17. When we sorted the movies by year using the .head() are Gone with the Wind, Bambi, 101 Dalmatians, The Jungle Book, The Godfather

#### End of Lab 06